kube-core/src/conversion.rs

```diff
@@ -0,0 +1,141 @@
+
+use self::low_level::{ConversionResponse, ConversionReview};
+
+/// Defines low-level typings.
+pub mod low_level;
+/// High-level easy-to-use converter.
+mod star;
+
+pub use star::{StarConverter, StarConverterBuilder, StarRay};
+
+/// Helper which implements low-level API based on the provided converter.
+#[derive(Clone)]
+pub struct ConversionHandler {
+    converter: Arc<dyn Converter + Send + Sync>,
+}
+
+impl ConversionHandler {
+    /// Creates new `ConversionHandler` which will use given `converter` for operation.
+    pub fn new<C: Converter + Send + Sync + 'static>(converter: C) -> Self {
+        ConversionHandler {
+            converter: Arc::new(converter),
+        }
+    }
+
+    /// Processes `request` using stored converter so that you don't need
+    /// to think about copying `.uid` and other boring things.
+    /// Returned value is ready to be serialized and returned.
+    pub fn handle(&self, review: ConversionReview) -> ConversionReview {
+        let req = match review.request {
+            Some(r) => r,
+            None => {
+                return ConversionReview {
+                    types: review.types.clone(),
+                    request: None,
+                    response: Some(ConversionResponse::error(
+                        String::new(),
+                        ".request is unset in input".to_string(),
+                    )),
+                }
+            }
+        };
+
+        let mut converted_objects = Vec::new();
+        for (idx, object) in req.objects.into_iter().enumerate() {
+            match self.converter.convert(object, &req.desired_api_version) {
+                Ok(c) => converted_objects.push(c),
+                Err(error) => {
+                    let msg = format!("Conversion of object {} failed: {}", idx, error);
+                    return ConversionReview {
+                        types: review.types.clone(),
+                        request: None,
+                        response: Some(ConversionResponse::error(req.uid, msg)),
+                    };
+                }
+            }
+        }
+        ConversionReview {
+            types: review.types.clone(),
+            request: None,
+            response: Some(ConversionResponse::success(req.uid, converted_objects)),
+        }
+    }
+}
+
+/// Converter is entity which supports all `N*(N-1)` possible conversion directions.
+/// This trait does not specify strategy used. You may implement this trait yourself
+/// or use [`StarConverter`](StarConverter).
+pub trait Converter {
+    /// Actually performs conversion.
+    /// # Requirements
+    /// All metadata fields except labels and annotations must not be mutated.
+    /// # Errors
+    /// While the signature allows returning errors, it is discouraged.
+    /// Ideally, conversion should always succeed.
+    fn convert(
+        &self,
+        object: serde_json::Value,
+        desired_api_version: &str,
+    ) -> Result<serde_json::Value, String>;
```

Comment:
1. User will likely want to work with concrete type, so either way downcast is required.
2. There is no way to go from `serde_json::Value` to `dyn ...` without requiring additional information from user (i.e. we can't guess which `Deserialize::deserialize` function to call based on just api version and kind). And I see no "nice" way to ask this information.

I think that we may add a `ResourceRegistry` with the following API:
```rust
type ResourceRegistry;
/// Creates empty registry
fn new();
/// Adds resource to the registry.
fn register<K: Resource>(&mut self, dt: &K::DynamicType);
/// if apiVersion and kind are known, parses this obj and upcasts to Box<Any>, otherwise returns None
parse(&self, obj: serde_json::Value) -> Option<serde_json::Result<Box<dyn Any>>>
```
Or, alternatively, more complex and more type-safe API
```rust
type ResourceRegistry<U>; // when U = Box<dyn Any>, equivalent to the previous API

/// Creates empty registry
fn new();
/// Adds resource to the registry.
fn register<K: Resource + Into<U>>(&mut self, dt: &K::DynamicType);
/// if apiVersion and kind are known, parses this obj and upcasts to U, otherwise returns None
parse(&self, obj: serde_json::Value) -> Option<serde_json::Result<U>>
```

Now user can create such a registry and than use it inside `impl Conversion`, but it is also useful in other contexts (e.g. writing admission webhook for multiple resources, or re-implementing `kubectl describe` in rust).

Reply:
2. I think that is probably overkill for kube and what it is; Deserialisation into a specific type. If there are many types the conversion hook deals with (which i think is a rare case), then it should be up to the app to multiplex.

1. If we make the trait generic for the serialization type, won't this work?

```rust
pub trait Conversion<Source: DeserializeOwned, Destination: Serialize> {
    fn convert(&self, obj: Source, desired_version: &str) -> Result<Destination>;
}
```

if `Source` turns out to be `DynamicObject` then that's fine (easy to convert from a `Value`), and if users just convert them into their own `K` then  we do deserialization early and have this fn focus on the business logic - which it feels like it should be doing.

..probably would be hard to turn it into a generic `TryFrom` but that would be ideal.

Reply:
In current design, `Conversion` represents not one possible conversion, but all of them (i.e. it does one of the `N*(N-1)` things depending on input and desired versions, where N is total version count). Consequently `Source = Destination` (because each version may appear as both input and output). And this is either `DynamicObject` (or other highly dynamic value), or `serde(untagged)`-enum of all supported versions.
So some downstream converter may look like
```rust
#[serde(untagged)]
enum FooVersions {
    V1(FooV1),
    V2(FooV2),
}

struct FooConversion;

impl Conversion<FooVersions> for FooConversion {
    fn convert(&self, input: FooVersions, desired_version: &str) {
         match (input, desired_version) {
             (V1(v1), "v2") => /*convert v1 -> v2*/,
             (V2(v2), "v1") => /*convert v2 -> v1*/,
             _ => /*noop conversions v1 -> v1, or v2 -> v2*/
         }
    }
}
```
Better than parsing `serde_json::Value`-es, but still seems like a boilerplate.

Alternatively, we may change `Conversion` responsibility - from "thing which converts any version into any other" to "thing which converts from one particular version into another particular version". This will require putting more type-erasure magic into `ConversionHandler`, but that's probably worth it.
This api usage might look something like:
```
impl Conversion<FooV1, Foov2> for FooConversion {
   // strongly-typed!
   fn convert(&self, foo: FooV1) -> Result<FooV2, String>;
}

impl Conversion<FooV2, FooV1> foo FooConversion {
  // strongly-typed!
  fn convert(&self, foo: FooV2) -> Result<FooV1>, String;
}

let handler = ConversionHandlerBuilder::new(FooConversion {})
    .register::<FooV1, FooV2>()
    .register::<FooV2, FooV1>()
    .build();
```

What do you think?

> probably would be hard to turn it into a generic TryFrom

Yep, I don't see a way to go from `impl TryFrom<FooV1> for FooV2, impl TryFrom<FooV2> for V1` to `FooVersions + string -> FooVersions`.

Reply:
Ah, I see. Yeah I think it would be better if the user just wrote plain conversions from type1 to type2 rather than a match (particularly if the normal case is a single in-flight version migration) and lift the version->type matching into the handler somehow.

In your expression:

> impl TryFrom<FooV1> for FooV2, impl TryFrom<FooV2> for V1 to FooVersions + string -> FooVersions

What is the string there? The desired version?

I imagine if all the types we register `impl Resource` we wouldn't even need to tell the handler what version it is, it can figure that out via `api_version()` and then deserialize into the correct one at runtime.

Reply:
> What is the string there? The desired version?

Yes, from the Kubernetes PoV conversion webhook is `fn(obj: json, desired: string) -> converted: json`.

>  lift the version->type matching into the handler somehow.

OK, I'm sure we can achieve this.